DIPs/DIP1004.md

```diff
@@ -0,0 +1,272 @@
+    }
+}
+
+void main ( )
+{
+    auto d1 = new Derived();  // ok
+    assert(d1.sum == 10);
+
+    auto d2 = new Derived(10, 10);  // ok
+    assert(d2.sum == 20);
+}
+```
+
+## Proposed features
+
+### 1. Add implicit constructor inheritance for derived classes with no constructors
+
+If a derived class does not define any constructors of its own then that class
+will automatically inherit all base class constructors.
+
+Example:
+
+```D
+class ParseException : Exception
+{
+    // inherits all of Exception's constructors
+}
+
+void parseArgs ( string[] args )
+{
+    if (args < 2)
+    {
+        throw new ParseException("Expected at least one argument");
+    }
+}
+
+int main ( string[] args )
+{
+    try
+    {
+        parseArgs(args);
+
+        return 0;
+    }
+    catch (ParseException ex)
+    {
+        stderr.writefln("Parsing error: %s", ex.msg);
+    }
+    catch (Exception ex)
+    {
+        stderr.writefln("Unexpected error: %s", ex.msg);
+    }
+
+    return 1;
+}
+```
+
+**Rationale**: If a derived class does not define any of its own constructors then
+there is no danger of leaving the object in an unexpected state.
+
+Side-note: Forwarding constructors are most prominent in exception class hierarchies.
+
+### 2. Add language support for inheriting base class constructors for derived classes with constructors
+
+If the author of the derived class defines its own set of constructors but also
+wants to allow construction via the base class'es constructors then the
+new constructor inheritance syntax can be used for this purpose:
+
+```D
+enum ErrorCode
+{
+    FileNotFound
+}
+
+class FileException : Exception
+{
+    /// introduce overload which takes an error code
+    this ( ErrorCode error_code, string file = __FILE__, string line = __LINE__ )
+    {
+        super(error_code.to!string, file, line);
+    }
+
+    /// inherit base class constructors
+    alias super.this this;
+}
+
+void main ( string[] args )
+{
+    // using Exception's constructor
+    throw new FileException("Something went wrong");
+
+    // using FileException's constructor
+    throw new FileException(ErrorCode.FileNotFound);
+}
+```
+
+This feature is similar to D's existing feature for re-introducing base class
+methods to allow overloading of a base class method:
+
+```D
+class Base
+{
+    int sum ( int x, int y ) { return x + y; }
+}
+
+class Derived : Base
+{
+    // re-introduces base class method to allow overloading
+    alias super.sum sum;
+
+    // overload of the base class sum method
+    float sum ( float x, float y ) { return x + y; }
+}
+
+void main ( string[] args )
+{
+    auto d = new Derived;
+    d.sum(int(1), int(2));
+    d.sum(float(1), float(2));
+}
+```
+
+## Breaking changes
+
+There are no expected breaking changes with this proposal.
+
+The feature syntax shares a similarity with object subtyping,
+however `alias this.super this` would not conflict with the subtyping feature
+since `super` is a reserved word.
```

Comment:
Hmm I haven't, I did assume the compiler would currently reject `alias this.super this`.